dev/src/bulk-writer.ts

```diff
@@ -0,0 +1,428 @@
+   * updated by providing dot-separated field path strings or by providing
+   * FieldPath objects.
+   *
+   * A Precondition restricting this update can be specified as the last
+   * argument.
+   *
+   * @param {DocumentReference} documentRef A reference to the document to be
+   * updated.
+   * @param {UpdateData|string|FieldPath} dataOrField An object containing the
+   * fields and values with which to update the document or the path of the
+   * first field to update.
+   * @param {...(Precondition|*|string|FieldPath)} preconditionOrValues - An
+   * alternating list of field paths and values to update or a Precondition to
+   * restrict this update
+   * @returns {Promise<WriteResult>} A promise that resolves with the result of
+   * the write. Throws an error if the write fails.
+   *
+   *
+   * @example
+   * let bulkWriter = firestore.bulkWriter();
+   * let documentRef = firestore.collection('col').doc();
+   *
+   * bulkWriter
+   *  .update(documentRef, {foo: 'bar'})
+   *  .then(result => {
+   *    console.log('Successfully executed write at: ', result);
+   *  })
+   *  .catch(err => {
+   *    console.log('Write failed with: ', err);
+   *  });
+   * });
+   */
+  update(
+    documentRef: DocumentReference,
+    dataOrField: UpdateData | string | FieldPath,
+    ...preconditionOrValues: Array<
+      {lastUpdateTime?: Timestamp} | unknown | string | FieldPath
+    >
+  ): Promise<WriteResult> {
+    const bulkCommitBatch = this.getEligibleBatch(documentRef);
+    bulkCommitBatch.refsInBatch.add(documentRef);
+    bulkCommitBatch.writeBatch.update(
+      documentRef,
+      dataOrField,
+      preconditionOrValues
+    );
+
+    return this.getResult(
+      bulkCommitBatch.tag,
+      bulkCommitBatch.writeBatch.opCount
+    );
+  }
+
+  /**
+   * Returns a Promise that resolves when there are no more pending writes. It
+   * never fails.
+   *
+   * The promise resolves immediately if there no pending writes. Otherwise, the
+   * Promise waits for all previously issued writes, but it does not wait for
+   * writes that were added after the method is called. If you want to wait for
+   * additional writes, call `waitForPendingWrites()` again.
+   *
+   * @return {Promise<void>} A promise that resolves when there are no more
+   * pending writes.
+   *
+   * @example
+   * let bulkWriter = firestore.bulkWriter();
+   *
+   * bulkWriter.create(documentRef, {foo: 'bar'});
+   * bulkWriter.update(documentRef2, {foo: 'bar'});
+   * bulkWriter.delete(documentRef3);
+   * await waitForPendingWrites().then(() => {
+   *   console.log('Successfully executed all writes');
+   * });
+   */
+  async waitForPendingWrites(): Promise<void> {
+    const trackedBatches = this.batchQueue;
+    this.sendBatches();
+    const writePromises = trackedBatches.map(batch => {
+      batch.pending = true;
+      return this.resultsMap.get(batch.tag)!.promise;
+    });
+    await Promise.all(writePromises);
```

Comment:
Nit: If you are using `await` only in the last line, you should just use `return`.